### PR TITLE
Fix the warnings and problems Qodana found

### PR DESCRIPTION
**Pull request by NealGeilen**

### CODE REVIEW

File: Epsilon.Abstractions/OutcomeGradeStatus.cs [hunk 1]
```diff
@@ -1,5 +1,8 @@
+using System.Diagnostics.CodeAnalysis;
+
 namespace Epsilon.Abstractions;
 
+[SuppressMessage("ReSharper", "UnusedType.Global")]
```
Comment (Typiqally) [inline Epsilon.Abstractions/OutcomeGradeStatus.cs:5]: Could you add justification to this, smth like will be implemented in the near future? Just for clarity

File: Epsilon.Canvas.Abstractions/Rest/Page.cs [hunk 2]
```diff
@@ -31,13 +31,13 @@ public Page(string id)
     public string? CreatedAt { get; init; }
 
     [JsonPropertyName("url")]
-    public string? Url { get; init; }
+    private string? Url { get; }
```
Comment (Typiqally) [inline Epsilon.Canvas.Abstractions/Rest/Page.cs:34]: I don't think these should be private, could you give a reason?

File: Epsilon.Host.WebApi/Program.cs [hunk 3]
```diff
@@ -84,7 +84,7 @@
     app.UseCors(static x => x
                             .AllowAnyMethod()
                             .AllowAnyHeader()
-                            .SetIsOriginAllowed(static origin => true)
+                            .SetIsOriginAllowed(static origin => origin.Length > 0)
                             .AllowCredentials());
```
Comment (Typiqally) [inline Epsilon.Host.WebApi/Program.cs:88]: We shouldn't rly add code like this, it's just bad practise in general.

File: Epsilon/Services/LearningOutcomeCanvasResultService.cs [hunk 4]
```diff
@@ -68,15 +68,18 @@ public async IAsyncEnumerable<LearningDomainSubmission> GetSubmissions(string st
 
         await Task.WhenAll(submissionsTask, domainOutcomesTask);
 
-        foreach (var submission in submissionsTask?.Result?.Courses?.SelectMany(static c => c.Submissions.Nodes))
+        if (submissionsTask.Result?.Courses != null)
```
Comment (Typiqally) [inline Epsilon/Services/LearningOutcomeCanvasResultService.cs:71]: I think this is good, but to reduce nesting maybe you can inverse the if and throw an exception or something like that. It's also better for debugging and error handling

File: Epsilon/Services/LearningOutcomeCanvasResultService.cs [hunk 5]
```diff
@@ -97,36 +100,43 @@ private static IEnumerable<LearningDomainCriteria> GetSubmissionCriteria(Submiss
...
+
     private static IEnumerable<LearningDomainOutcomeRecord> GetOutcomeResults(
         Submission submission,
         Task<IEnumerable<LearningDomainOutcome?>>? domainOutcomesTask
     )
     {
         var outcomeRecords = new List<LearningDomainOutcomeRecord>();
-        //Loop trough all submissions of assignment to look for outcomes that have been graded. 
-        foreach (var submissionHistory in submission.SubmissionHistories.Nodes.OrderByDescending(static s => s.SubmittedAt))
+        if (submission.SubmissionHistories?.Nodes != null)
```
Comment (Typiqally) [inline Epsilon/Services/LearningOutcomeCanvasResultService.cs:110]: Same as previous comment

File: qodana.yaml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 version: 1.0
-failThreshold: 64
+failThreshold: 0
```
Comment (Typiqally) [inline qodana.yaml:2]: Do you know if this also picks up on info comments from Qodana? Just a question.

**Review by Typiqally — CHANGES_REQUESTED** [review]

File: qodana.yaml (hunk 6 above)
Comment (NealGeilen) [reply qodana.yaml:2]: No, It counts 'problems' as in notices, warnings, and errors

File: Epsilon.Canvas.Abstractions/Rest/Page.cs (hunk 2 above)
Comment (NealGeilen) [reply Epsilon.Canvas.Abstractions/Rest/Page.cs:34]: Fair, But for now it is only used within the class. So this is an option or suppress the notice.

File: Epsilon.Abstractions/OutcomeGradeStatus.cs [hunk 7]
```diff
@@ -1,5 +1,9 @@
+using System.Diagnostics.CodeAnalysis;
+
 namespace Epsilon.Abstractions;
 
+//Suppressed for later implementation. Remove when implemented  
+[SuppressMessage("ReSharper", "UnusedType.Global")]
```
Comment (Typiqally) [inline Epsilon.Abstractions/OutcomeGradeStatus.cs:6]: You can also do it like this, which is the most common way: `[SuppressMessage("a", "b", Justification = "justify here")]`

File: Epsilon/Services/LearningOutcomeCanvasResultService.cs [hunk 8]
```diff
@@ -68,12 +69,17 @@ public async IAsyncEnumerable<LearningDomainSubmission> GetSubmissions(string st
 
         await Task.WhenAll(submissionsTask, domainOutcomesTask);
 
-        foreach (var submission in submissionsTask?.Result?.Courses?.SelectMany(static c => c.Submissions.Nodes))
+        if (submissionsTask.Result?.Courses == null)
+        {
+            throw new NoNullAllowedException();
```
Comment (Typiqally) [inline Epsilon/Services/LearningOutcomeCanvasResultService.cs:74]: Use this one and please add a message like such: `throw new NullReferenceException("msg here");`

**Review by Typiqally — CHANGES_REQUESTED** [review]

**Review by Typiqally — APPROVED** [review]

**Pull request merged**